app/javascript/mastodon/selectors/index.js

```diff
@@ -62,27 +62,41 @@ const regexFromFilters = filters => {
   }).join('|'), 'i');
 };
 
+const regexFromFilters = filters => {
+  return filters.map(filter => [regexFromKeywords(filter.get('keywords')), filter.get('title')]);
+};
+
 // Memoize the filter regexps for each valid server contextType
 const makeGetFiltersRegex = () => {
   let memo = {};
 
   return (state, { contextType }) => {
-    if (!contextType) return ImmutableList();
+    if (!contextType) return [null, null];
 
     const serverSideType = toServerSideType(contextType);
-    const filters = state.get('filters', ImmutableList()).filter(filter => filter.get('context').includes(serverSideType) && (filter.get('expires_at') === null || Date.parse(filter.get('expires_at')) > (new Date())));
+    const now = new Date();
+    const filters = state.get('filters', ImmutableMap()).toList().filter(filter => filter.get('context').includes(serverSideType) && filter.get('keywords') && filter.get('keywords').size > 0 && (filter.get('expires_at') === null || filter.get('expires_at') > now));
```

Comment:
As I understand, server-side filtering logic is still re-implemented client-side to handle filter expiration, thereby reducing the benefits of the new implementation. Ideally, the client would not need to construct its own regexes at all. Why can't the server-side match in the API response include an `expires_at` property, and then the client can check *that*, but does not have to re-match any filter regexes?

Reply:
The API response includes the `expires_at`, but the whole client-side reimplementation is not for that purpose. It's for when filters get changed (new filters added require re-filtering everything against the added filters, and keywords changing require re-filtering everything against the updated filters).

Reply:
Yes, a page refresh is enough. I'll see how much simpler the WebUI code gets by removing this.

Reply:
I'm not too sure about dropping this behavior, but I pushed a commit anyway, so you can judge whether the simplification is worth it.